### PR TITLE
RSDP fixes

### PR DESCRIPTION
**Pull request by sameo**

### CODE REVIEW

File: hw/arm/virt-acpi-build.c [hunk 1]
```diff
@@ -388,10 +390,15 @@ build_rsdp(GArray *rsdp_table, BIOSLinker *linker, unsigned xsdt_tbl_offset)
         ACPI_BUILD_RSDP_FILE, xsdt_pa_offset, xsdt_pa_size,
         ACPI_BUILD_TABLE_FILE, xsdt_tbl_offset);
 
-    /* Checksum to be filled by Guest linker */
+    /* Legacy checksum to be filled by Guest linker */
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: per spec it's 'Checksum' so leave it as is

File: hw/arm/virt-acpi-build.c [hunk 2]
```diff
@@ -388,10 +390,15 @@ build_rsdp(GArray *rsdp_table, BIOSLinker *linker, unsigned xsdt_tbl_offset)
...
     bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE,
-        (char *)rsdp - rsdp_table->data, sizeof *rsdp,
+        (char *)rsdp - rsdp_table->data, xsdt_offset,
         (char *)&rsdp->checksum - rsdp_table->data);
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: I'd use explicit sizes to avoid pointer arithmeics

File: hw/arm/virt-acpi-build.c [hunk 3]
```diff
@@ -388,10 +390,15 @@ build_rsdp(GArray *rsdp_table, BIOSLinker *linker, unsigned xsdt_tbl_offset)
...
+
+    /* Extended checksum to be filled by Guest linker */
+    bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE,
+        (char *)rsdp - rsdp_table->data, sizeof *rsdp,
+        (char *)&rsdp->extended_checksum - rsdp_table->data);
 }
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: I'm obviosly biased here but I'd suggest to pull in and reuse 2 commits from
 https://github.com/imammedo/qemu/commits/rsdp_fixes
as commit messages and bodies are more precisely describe what's going on

File: hw/arm/virt-acpi-build.c [hunk 4]
```diff
@@ -401,6 +401,17 @@ build_rsdp(GArray *rsdp_table, BIOSLinker *linker, unsigned xsdt_tbl_offset)
         (char *)&rsdp->extended_checksum - rsdp_table->data);
 }
 
+static void
+init_rsdp_data(AcpiRsdpData *data, char *oem_id, uint8_t revision,
+               unsigned *rsdt_offset, unsigned *xsdt_offset)
+{
+    memcpy(data->oem_id,
+           (oem_id && strlen(oem_id) >= 6) ? oem_id : ACPI_BUILD_APPNAME6, 6);
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: I'd drop conditional here and make user always provide oem_id

File: hw/arm/virt-acpi-build.c [hunk 5]
```diff
@@ -859,7 +871,8 @@ void virt_acpi_build(VirtMachineState *vms, AcpiBuildTables *tables)
     build_xsdt(tables_blob, tables->linker, table_offsets, NULL, NULL);
 
     /* RSDP is in FSEG memory, so allocate it separately */
-    build_rsdp(tables->rsdp, tables->linker, xsdt);
+    init_rsdp_data(&rsdp, NULL, ACPI_RSDP_REV_2, NULL, &xsdt);
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: maybe drop init function here and just declare at the begining od acpi_build() an AcpiRsdpData variable  with initializer, like its done in init_common_fadt_data() and then assigh xsdt offset before callin build_rsdp()

File: hw/arm/virt-acpi-build.c [hunk 6]
```diff
@@ -368,37 +368,47 @@ static void acpi_dsdt_add_power_button(Aml *scope)
...
-    memcpy(&rsdp->signature, "RSD PTR ", sizeof(rsdp->signature));
-    memcpy(rsdp->oem_id, rsdp_data->oem_id, sizeof(rsdp->oem_id));
-    rsdp->length = cpu_to_le32(sizeof(*rsdp));
-    rsdp->revision = rsdp_data->revision;
+    /* RSDP signature */
+    g_array_append_vals(tbl, "RSD PTR ", 8);
 
-    /* Address to be filled by Guest linker */
+    /* Space for the legacy checksum */
+    build_append_int_noprefix(tbl, 0, 1);
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: move comment after function if it fits in 80c.
Also  preferable a table comment should be just field name from spec without extra info, or at least be the first thing in it if necessary.

File: hw/arm/virt-acpi-build.c [hunk 7]
```diff
@@ -368,37 +368,47 @@ static void acpi_dsdt_add_power_button(Aml *scope)
...
+
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: drop empty lines between fields, they don't help. Let the table block be as a whole

File: hw/arm/virt-acpi-build.c [hunk 8]
```diff
@@ -368,37 +368,47 @@ static void acpi_dsdt_add_power_button(Aml *scope)
...
+    build_append_int_noprefix(tbl, rsdp_data->revision, 1);
+
+    /* Space for the RSDT address (32 bit) */
+    build_append_int_noprefix(tbl, 0, 4);
+
+    /* Length */
+    build_append_int_noprefix(tbl, 36, 4);
+
+    /* XSDT address to be filled by guest linker */
+    build_append_int_noprefix(tbl, 0, 8); /* XSDT address (64 bit) */
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: drop ' (64 bit)'

File: hw/arm/virt-acpi-build.c [hunk 9]
```diff
@@ -368,37 +368,47 @@ static void acpi_dsdt_add_power_button(Aml *scope)
...
+
+    /* Space for the reserved bytes */
+    build_append_int_noprefix(tbl, 0, 3);
 
-    /* Legacy checksum to be filled by Guest linker */
-    bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE,
-        (char *)rsdp - rsdp_table->data, xsdt_offset,
-        (char *)&rsdp->checksum - rsdp_table->data);
+    /* Legacy checksum to be filled by guest linker */
+    bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE, 0, 20, 8);
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: it's either intermix linker with  fields or move all linker commands to the end if flow allows.
So move  bios_linker_loader_add_checksum right after a corresponding field or do it all at the end.

File: hw/arm/virt-acpi-build.c [hunk 10]
```diff
@@ -368,37 +368,47 @@ static void acpi_dsdt_add_power_button(Aml *scope)
...
 
     /* Extended checksum to be filled by Guest linker */
-    bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE,
-        (char *)rsdp - rsdp_table->data, sizeof *rsdp,
-        (char *)&rsdp->extended_checksum - rsdp_table->data);
+    bios_linker_loader_add_checksum(linker, ACPI_BUILD_RSDP_FILE, 0, 36, 32);
 }
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: minor comments aside conversion looks good

File: hw/arm/virt-acpi-build.c [hunk 11]
```diff
@@ -388,27 +388,41 @@ build_rsdp(GArray *tbl, BIOSLinker *linker, AcpiRsdpData *rsdp_data)
     /* Space for the RSDT address (32 bit) */
     build_append_int_noprefix(tbl, 0, 4);
 
-    /* Length */
-    build_append_int_noprefix(tbl, 36, 4);
+    if (rsdp_data->revision == ACPI_RSDP_REV_1 &&
+        rsdp_data->rsdt_tbl_offset) { /* Legacy RSDP */
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: if rev1 then rsdt is must, so g_assert(rsdp_data->rsdt_tbl_offset) would be better

File: hw/arm/virt-acpi-build.c [hunk 12]
```diff
@@ -388,27 +388,41 @@ build_rsdp(GArray *tbl, BIOSLinker *linker, AcpiRsdpData *rsdp_data)
...
+                /* RSDT address to be filled by guest linker */
+                bios_linker_loader_add_pointer(linker,
```
Comment (imammedo) [inline hw/arm/virt-acpi-build.c]: potentially there could be both rsdt and xsdt pointers, so maybe use following approach

if(rsdt != NULL)
      add_pointer()
if (rev1)
     return;
rev2 fields
return

Also this way code is less indented, that helps with readability

File: hw/acpi/aml-build.c [hunk 13]
```diff
@@ -1589,6 +1589,76 @@ void acpi_build_tables_cleanup(AcpiBuildTables *tables, bool mfre)
     g_array_free(tables->vmgenid, mfre);
 }
```
Comment (imammedo) [inline hw/acpi/aml-build.c:1591]: comment to commit message:
My English is rather bad but maybe do something like this:
s/we can export it .../move it to generic location and reuse it in x86 code, to reduce code duplication/

File: include/hw/acpi/acpi-defs.h [hunk 14]
```diff
@@ -40,19 +40,6 @@ enum {
...
-    uint8_t  oem_id [6];             /* OEM identification */
-    uint8_t  revision;               /* Must be 0 for 1.0, 2 for 2.0 */
-    uint32_t rsdt_physical_address;  /* 32-bit physical address of RSDT */
-    uint32_t length;                 /* XSDT Length in bytes including hdr */
-    uint64_t xsdt_physical_address;  /* 64-bit physical address of XSDT */
-    uint8_t  extended_checksum;      /* Checksum of entire table */
-    uint8_t  reserved [3];           /* Reserved field must be 0 */
-} QEMU_PACKED;
-typedef struct AcpiRsdpDescriptor AcpiRsdpDescriptor;
-
```
Comment (imammedo) [inline include/hw/acpi/acpi-defs.h]: squash it in previous patch and mention that in commit message

**Comment by sameo** [comment]
v1 sent to the QEMU ml:
https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg05197.html

**Comment by sameo** [comment]
v2: https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg06190.html

**Comment by rbradford** [comment]
This project is no longer active and being archived. Cloud Hypervisor is the successor: https://github.com/cloud-hypervisor/cloud-hypervisor

**Pull request closed without merging**